### PR TITLE
[#3186] Manage Access - Vue Instance

### PR DESCRIPTION
**Pull request by Maurier**
This PR follows up on #3186 by implementing a vue instance containing all frontend logic related to Manage Access. 

With this work we start seeing the benefits of Vue.js. The code has been simplified significantly and we should see a more reactive UI. All functionality inside the Manage Access modal is now performed via AJAX and no page refresh is required.

Notice that changing resource owners won't yet reflect on the Owners listing on the top left of the page. This will be implemented later on when rewriting logic for this "Highlights" component.

However, changing sharing status does now reflect on other items on the page because this was easy to setup temporarily until the "Highlights" component is implemented.

This work can be tested at https://riverdev.hydroshare.org

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [ ] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. Create a resource
2. Test functionality inside Manage Access
3. Verify that no action refreshes the page and the UI state is handled correctly.

### CODE REVIEW

**Comment by Maurier** [comment]
@cuahsimarko, @pkdash - This is now ready for review. Note that I changed some responses in the backend to handle AJAX requests.

File: hs_core/models.py [hunk 1]
```diff
@@ -250,27 +309,45 @@ def page_permissions_page_processor(request, page):
...
+        "self_access_level": self_access_level,
         "can_change_resource_flags": can_change_resource_flags,
         "is_replaced_by": is_replaced_by,
         "is_version_of": is_version_of,
         "show_manage_access": show_manage_access
     }
 
 
+# TODO: import this from hydroshare_tags
+def best_name(content):
```
Comment (Maurier) [inline hs_core/models.py]: @cuahsimarko - I was unnable to import this method from hydroshare_tags. Perhaps you could give it a try.

Comment (cuahsimarko) [reply hs_core/models.py]: When the "Dashboard" was pushed out, a new hs_docker_image was built. Please work with Scott or Phuong to engage with the 1.9.8 base image.

Comment (Maurier) [reply hs_core/models.py]: Not at all, I don't think this issue had to do with "timeago" or outdated image.

Comment (cuahsimarko) [reply hs_core/models.py]: After discussing with you, I'm still not sure why the import statement you tried isn't working. @sblack-usu is anything jumping out at you why from hs_core... import best_name isn't working for Mauriel here?

File: hs_core/models.py [hunk 2]
```diff
@@ -238,6 +237,66 @@ def page_permissions_page_processor(request, page):
         for edit_grp in edit_groups:
             edit_grp.can_undo = False
 
+    users_json = []
+
+    # TODO: simplify these iterations
+    for usr in owners:
```
Comment (cuahsimarko) [inline hs_core/models.py]: There are possibilities here, using dictionary/list comprehensions, but they promise to be more clever than readable. What you have here is probably 'least evil', despite the repetition. It is more readable than some alternatives I can imagine.

File: theme/static/js/collection-resource.js [hunk 3]
```diff
@@ -204,15 +204,16 @@ function remove_collection_item_ajax(res_id, move_to_candidate_list) {
                 if (json_response.hasOwnProperty('metadata_status')) {
                     if (json_response.metadata_status !== $('#metadata-status').text()) {
                         $('#metadata-status').text(json_response.metadata_status);
-                        if (json_response.metadata_status.toLowerCase().indexOf("insufficient") == -1) {
+                        if (son_response.metadata_status.toLowerCase().indexOf("insufficient") == -1) {
                             var res_is_discoverable = $("#discoverable").val();
```
Comment (cuahsimarko) [inline theme/static/js/collection-resource.js]: json_response...

File: hs_core/views/__init__.py [hunk 4]
```diff
@@ -994,7 +1002,7 @@ def undo_share_resource_with_user(request, shortkey, user_id, *args, **kwargs):
         if undo_user_privilege == PrivilegeCodes.VIEW:
             undo_user_privilege = "view"
```
Comment (cuahsimarko) [inline hs_core/views/__init__.py]: I know it's not your code, but was === cleaner/more appropriate in these conditionals?

File: hs_core/views/__init__.py [hunk 5]
```diff
@@ -933,7 +941,7 @@ def _share_resource(request, shortkey, privilege, user_or_group_id, user_or_grou
                               'error_msg': err_message}
 
     else:
-        group_pic_url = 'No picture provided'
+        group_pic_url = None
         if group_to_share_with.gaccess.picture:
```
Comment (cuahsimarko) [inline hs_core/views/__init__.py]: did you test with and without to ensure it is ok with the None type instead of str?

File: hs_core/views/__init__.py [hunk 6]
```diff
@@ -96,11 +96,16 @@ def verify(request, *args, **kwargs):
 
 def change_quota_holder(request, shortkey):
     new_holder_uname = request.POST.get('new_holder_username', '')
```
Comment (cuahsimarko) [inline hs_core/views/__init__.py]: did you quickly search entire codebase for usages of change_quota_holder to ensure you know everything being impacted?

File: theme/static/js/hs-vue/subjet-keywords.js [hunk 7]
```diff
@@ -58,12 +58,12 @@ let subjKeywordsCmp = new Vue({
               {value: newVal.join(",")}, function (resp) {
                   if (resp.status === "success") {
                       vue.resKeywords = newVal;
+                      if (!newVal.length) {
+                          // If no keywords, the metadata is no longer sufficient to make the resource public
```
Comment (cuahsimarko) [inline theme/static/js/hs-vue/subjet-keywords.js]: what is the best practice in JS/jQuery/Vue for testing for undefined, null, empty string etc... guessing what you're doing here is fine

File: hs_core/models.py [hunk 8]
```diff
@@ -238,6 +237,66 @@ def page_permissions_page_processor(request, page):
...
+            "access": "view",
+            "id": usr.id,
+            "pictureUrl": usr.gaccess.picture.url if usr.gaccess.picture else None,
+            "best_name": usr.name,
+            "user_name": None,
+            "can_undo": usr.can_undo
+        })
+
+    users_json = json.dumps(users_json)
+
```
Comment (cuahsimarko) [inline hs_core/models.py]: seems like the Python `usr` User object should have an `access` field... that would be another PR likely, though to refactor all that and change necessary Models if that applies

Comment (cuahsimarko) [reply hs_core/models.py]: then, there wouldn't be separate Lists for Users by access type... and mapping from the Lists to the dict for JSON serialization would be a dictionary comprehension

File: hs_core/models.py (hunk 1 above)
Comment (sblack-usu) [reply hs_core/models.py]: I was able to import without any trouble, what does your import look like?  It should be something like this
`from .templatetags.hydroshare_tags import best_name`

File: theme/static/js/collection-resource.js (hunk 3 above)
Comment (cuahsimarko) [reply theme/static/js/collection-resource.js]: this is a typo, so must fix

File: theme/templates/includes/access-row.html [hunk 9]
```diff
@@ -1,133 +1,89 @@
...
 
-{% for u in entries %}
-<tr id="row-id-{{ u.pk }}"
-    {% if access != "Is owner" %}
-    {% elif entries|length == 1 or not is_owner_user or u.pk == quota_holder.pk %}
-    class="hide-actions"
-    {% endif %}>
+<tr v-for="(user, index) in this.users"
+    :class="{ 'hide-actions': selfAccessLevel !== 'owner' && (user.access == 'owner' && hasOnlyOneOwner || user.id === quotaHolder.id), loading: user.loading}">
     <td>
```
Comment (cuahsimarko) [inline theme/templates/includes/access-row.html]: I still get confused when to use double or triple = between user.access and 'owner' do you mean for three?

File: hs_core/views/__init__.py (hunk 5 above)
Comment (Maurier) [reply hs_core/views/__init__.py]: Yes, I tested this both with and without a picture.

File: theme/templates/includes/access-row.html [hunk 10]
```diff
@@ -1,133 +1,89 @@
...
         <table class="user-scope">
-            {% if access_type == "user" %}
-            <tr>
+            <tr v-if="user.user_type === 'user'">
                 <td>
```
Comment (cuahsimarko) [inline theme/templates/includes/access-row.html]: for here and all your v-if v-<conditionals> did you ensure you thought through 'truthy' state for all possible outcomes undefined, empty string, etc.?

File: hs_core/views/__init__.py (hunk 4 above)
Comment (Maurier) [reply hs_core/views/__init__.py]: Using `===`  just means it will also check for the type. So, yes. It is almost always better to use it.

File: theme/templates/includes/access-row.html [hunk 11]
```diff
@@ -1,133 +1,89 @@
...
-                    <div data-col="name" class="user-name">
-                        <a href="/user/{{ u.id }}">{{ u|best_name }}</a>
+                    <div class="user-name">
+                        <a :href="'/user/' + user.id">${user.best_name}</a>
                     </div>
-                    <div data-col="user-name"
-                          class="user-username-content">{{ u.username }}</div>
+                    <div class="user-username-content">${user.user_name}</div>
                     <div>
```
Comment (cuahsimarko) [inline theme/templates/includes/access-row.html]: Ok I think you set up a custom handler for $... is another valid (but less readable) approach to use the django tags?

{% templatetag openvariable %} user.user_name {% templatetag closevariable %} 

when jQuery is present, I try to avoid hijacking the $ symbol...

File: hs_core/views/__init__.py (hunk 6 above)
Comment (Maurier) [reply hs_core/views/__init__.py]: Yes, the only other usages I saw were in some tests that I am now updating.

File: theme/static/js/hs-vue/subjet-keywords.js (hunk 7 above)
Comment (Maurier) [reply theme/static/js/hs-vue/subjet-keywords.js]: That is always a bit difficult to memorize, but I do keep in mind some conditional rules that I often use. Like empty string, zero, etc being equivalent to `false` when used as conditionals.

File: theme/static/js/hs-vue/manage-access.js [hunk 12]
```diff
@@ -0,0 +1,461 @@
+/**
+ * Created by Mauriel on 5/19/2019.
+ */
+
+let manageAccessCmp = new Vue({
+    el: '#manage-access',
+    delimiters: ['${', '}'],
```
[the remark this answers is not in the document]
Comment (Maurier) [reply theme/static/js/hs-vue/manage-access.js]: I agree. I will change this name. However, I will introduce some actual Vue components soon that will probably need to end with '-cmp'.

File: theme/templates/includes/access-row.html (hunk 9 above)
Comment (Maurier) [reply theme/templates/includes/access-row.html]: I do. I will make this change. Generally is good to use '===' because it will also check for the type.

File: theme/templates/includes/access-row.html (hunk 10 above)
Comment (Maurier) [reply theme/templates/includes/access-row.html]: I think those edge cases are not possible with our current API, so there are no edge cases handlers here.

File: theme/templates/includes/access-row.html (hunk 11 above)
Comment (Maurier) [reply theme/templates/includes/access-row.html]: I avoided using django escape tags because we might want to keep django tags running to render specific content inside Vue templates.

Since this runs in the HTML and jQuery runs on the scripts, I don't think there will ever be a conflict.

However, we could define better delimiters so I'm open to suggestions. One that comes to mind is double brackets: `[[ some_expression ]]`

Comment (cuahsimarko) [reply theme/templates/includes/access-row.html]: In that case, the way you have originally designed it is probably best.

File: theme/static/js/hs-vue/manage-access.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (cuahsimarko) [reply theme/static/js/hs-vue/manage-access.js]: I see. Also, we should look into .vue files for templates... we should be aware of all the MS Edge and IE10/11 issues, as well. Neal and I looked quickly at the compat matrix for Vue and it looked great, though overall

[the remark this answers is not in the document]
Comment (Maurier) [reply theme/static/js/hs-vue/manage-access.js]: Correct me if I'm wrong, but wouldn't we need full Djang-Vue.js integration in order to use .vue files? 

I'm not sure how challenging that would be. Here is a guide from which we could probably extrapolate what needs to be done:
https://github.com/michaelbukachi/django-vuejs-tutorial/wiki/Django-Vue.js-Integration-Tutorial

File: hs_core/views/__init__.py (hunk 4 above)
Comment (Maurier) [reply hs_core/views/__init__.py]: Oh, '===' are used in javaScript, so these ones do not to be updated.

Comment (cuahsimarko) [reply hs_core/views/__init__.py]: Yes, I was lost switching back and forth in the code review. GH view always messes with me.

Comment (cuahsimarko) [reply hs_core/views/__init__.py]: And earlier today I was at lunch on my phone, reading GH screens ....

File: hs_core/models.py (hunk 1 above)
Comment (Maurier) [reply hs_core/models.py]: Here is the error we are getting:
image

Were you able to import it in this file and running this branch?

Comment (sblack-usu) [reply hs_core/models.py]: So the problem is a circular import.  If you look at the last import in the stacktrace, you can see it's trying to make imports from hs_core.models, which is back to where you're stating the import for best_name.

You can get around this by limiting the scope of where you make this import by importing just before you need it, so line 241 seems like a good place to put this, `from templatetags.hydroshare_tags import best_name`

**Comment by Maurier** [comment]
@sblack-usu, @pkdash, @cuahsimarko I have now fixed the import issue following @sblack-usu's instructions. Please add your reviews when you get a chance so we can move this work forward. There are still a lot of things to do that depend on this one.

File: hs_core/templates/pages/collaborate.html [hunk 13]
```diff
@@ -89,7 +89,7 @@ <h4 class="modal-title">Members</h4>
                                                             <tbody>
                                                             {% for u in group.gaccess.members %}
                                                                 <tr id="row-id-{{ u.pk }}"
-                                                                    {% if owners|length == 1 or not is_owner_user %}class="hide-actions"{% endif %}>
+                                                                    {% if owners|length == 1 or not self_access_level == 'owner' %}class="hide-actions"{% endif %}>
```
Comment (cuahsimarko) [inline hs_core/templates/pages/collaborate.html]: is it best to be using classes, instead of v-if?

File: hs_core/templates/pages/genericresource.html [hunk 14]
```diff
@@ -830,8 +831,22 @@ <h3>Resource Specific</h3>
       const CURRENT_USER_ID = {{ request.user.id }};
       const SHORT_ID = "{{ cm.short_id }}";
       const STATIC_URL = "{{ STATIC_URL }}";
-      let spatial_coverage_type = "{{ spatial_coverage.type }}";
+      {% if show_manage_access %}
```
Comment (cuahsimarko) [inline hs_core/templates/pages/genericresource.html]: I don't recall coming across something like this before, with the Django template language, in the middle of a frontend <script> tag. Also, should this specify text/javascript or be more descriptive than just <script>?

File: hs_core/templates/pages/genericresource.html [hunk 15]
```diff
@@ -830,8 +831,22 @@ <h3>Resource Specific</h3>
...
+          const GROUP_IMAGE_DEFAULT_URL = "{{ STATIC_URL }}" + "img/home-page/step4.png";
+          const CAN_CHANGE_RESOURCE_FLAGS = "{{ can_change_resource_flags }}" === "True";
+          const CAN_BE_PUBLIC_OR_DISCOVERABLE = "{{ cm.can_be_public_or_discoverable }}" === "True";
+          const RESOURCE_ACCESS = {
+              isPublic: "{{ cm.raccess.public }}" === "True",
+              isDiscoverable: "{{ cm.raccess.discoverable }}" === "True",
+              isShareable: "{{ cm.raccess.shareable }}" === "True"
+          };
+      {% endif %}
+      var spatial_coverage_type = "{{ spatial_coverage.type }}";
```
Comment (cuahsimarko) [inline hs_core/templates/pages/genericresource.html]: Can conditional logic be moved to Vue or is that too complicated or even impossible here?

File: theme/templates/resource-landing-page/manage-access.html [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+                                    @click="isInviteUsers = true"
+                                    data-value="users">Users</button>
+                            <button type="button"
+                                    :class="{btn: true, 'btn-primary': !isInviteUsers, 'btn-default': isInviteUsers}"
+                                    @click="isInviteUsers = false"
+                                    data-value="groups">Groups</button>
+                        </div>
+
+                        <form id="add-access-form"
+                              :action="'/hsapi/_internal/' + resShortId + '/share-resource-with-' + (isInviteUsers ? 'user' : 'group') + '/'"
```
Comment (cuahsimarko) [inline theme/templates/resource-landing-page/manage-access.html]: Is there a way to have the JavaScript make the post, or will we always have a mix of forms/html POST and JS/Vue/jQuery POSTS? Just thinking big picture here.

Comment (cuahsimarko) [reply theme/templates/resource-landing-page/manage-access.html]: Consistency. I'm guessing we'll always have a mix.

**Comment by cuahsimarko** [comment]
All I was able to find was the close button bug, but not sure if that's present in develop. You should merge develop, now it has the release in it. The Close button on manage access bug could have been pre-existing/not in scope for this PR. I recommend a stakeholder/someone more experienced with the UI have to do a more detailed functional review of the UI. Some light comments in the source, left as well.

File: hs_core/models.py (hunk 2 above)
Comment (sblack-usu) [reply hs_core/models.py]: I think it's worth extracting the user json creation into it's own method and then calling the method 5 times while passing in the unique parameters.  It ensures we get the same format for each user type.

Comment (sblack-usu) [reply hs_core/models.py]: Do we know how this behaves with non-ascii characters in the user's names?

File: theme/templates/resource-landing-page/top-right-buttons.html [hunk 17]
```diff
@@ -11,7 +11,7 @@
             </a>
         {% endif %}
         {% if not resource_edit_mode %}
-            {% if not is_owner_user and not is_edit_user and not is_view_user and cm.raccess.public %}
+            {% if not self_access_level == 'None' and cm.raccess.public %}
                 {% if resource_is_mine %}
```
Comment (cuahsimarko) [inline theme/templates/resource-landing-page/top-right-buttons.html]: Is "Shareable" addressed here and elsewhere? I never fully understood that part. 
image

File: theme/templates/resource-landing-page/subject.html [hunk 18]
```diff
@@ -49,4 +52,4 @@ <h3>Subject Keywords</h3>
     </div>
 </div>
 
-<script type="text/javascript" src="{{ STATIC_URL }}js/hs-vue/subjet-keywords.js"></script>
```
Comment (cuahsimarko) [inline theme/templates/resource-landing-page/subject.html]: very minor but did you mean to name the file subject? The 'c' is missing.

File: theme/templates/resource-landing-page/subject.html [hunk 19]
```diff
@@ -33,7 +33,10 @@ <h3>Subject Keywords</h3>
                 </ul>
             </div>
             <div>
-                <p class="text-danger small" id="id-keywords-msg" v-if="showIsDuplicate"> Duplicate. Keyword not added.</p>
+                <p class="text-danger small" v-if="showIsDuplicate"> Duplicate. Keyword not added.</p>
+                <div v-if="error" class='alert alert-danger space-top small'>
+                    <strong>Error: </strong>${ error }
```
Comment (cuahsimarko) [inline theme/templates/resource-landing-page/subject.html]: Yes! v-if 💯

File: theme/templates/resource-landing-page/quota.html [hunk 20]
```diff
@@ -1,54 +1,59 @@
...
-            </span>
+<div v-if="selfAccessLevel === 'owner'" class="panel panel-default">
+    <div class="panel-heading">
+        <h3 class="panel-title">Storage</h3>
+    </div>
+    <div class="panel-body">
+        <span>The quota holder of this resource is: </span>
+        <span><a :href="'/user/' + quotaHolder.id">${quotaHolder.best_name}</a></span>
+        {# TODO: create a reusable component for this template #}
+{#        <span>#}
```
Comment (cuahsimarko) [inline theme/templates/resource-landing-page/quota.html]: clean up comments...

Comment (cuahsimarko) [reply theme/templates/resource-landing-page/quota.html]: I see, it's just a TODO so it's fine as-is

File: hs_core/templates/pages/collaborate.html (hunk 13 above)
Comment (sblack-usu) [reply hs_core/templates/pages/collaborate.html]: Are there any clear benefits to v-if? I'm concerned we're using two different tools to do the same thing.  I'd prefer not to do both unless there is a clear advantage to v-if.

File: hs_core/templates/pages/genericresource.html (hunk 14 above)
Comment (sblack-usu) [reply hs_core/templates/pages/genericresource.html]: I've seen it in this codebase before, though I don't have an opinion yet whether it's a good thing or not.  I don't have any ideas for alternative implementations of this.

Comment (sblack-usu) [reply hs_core/templates/pages/genericresource.html]: Looking into this more, I think it make sense to split these into some ajax calls when vue needs them instead of on the html.

Comment (Maurier) [reply hs_core/templates/pages/genericresource.html]: @sblack-usu - The advantage of having it this way is that Vue gets the data as soon as the template is passed to the client. Having ajax calls to get the data for each Vue instance would add unnecessary server load. 

@cuahsimarko - Django template language works in the entire .html file, even inside inline `script` tags. However, it would not work inside .js files which I think is what you mean.

File: hs_core/models.py (hunk 2 above)
Comment (Maurier) [reply hs_core/models.py]: I am not making changes to the way this data is parsed, so it should behave the same.

File: hs_core/templates/pages/collaborate.html (hunk 13 above)
Comment (Maurier) [reply hs_core/templates/pages/collaborate.html]: v-if is simpler and better because it removes the element from the DOM if the conditional is met (not just hiding it). Vue handles all of it and will add it back if the conditional dynamically changes.

File: hs_core/templates/pages/genericresource.html (hunk 15 above)
Comment (Maurier) [reply hs_core/templates/pages/genericresource.html]: I don't think it would be cleaner than this. Adding a JavaScript conditional here would imply still using a Django template tag to read `show_manage_access`  and then using it as a js conditional to declare the global constants inside the `If` statement scope (which is not recommended).

File: theme/templates/resource-landing-page/manage-access.html (hunk 16 above)
Comment (Maurier) [reply theme/templates/resource-landing-page/manage-access.html]: Actually, this form is not needed so I will be removing it. This post is now handled inside the Vue instance. I just forgot to remove it after moving the logic.

Comment (Maurier) [reply theme/templates/resource-landing-page/manage-access.html]: I think we'll still have a few forms in place, mainly for actions that redirect, thus not needing to be Vue-instanciated. For example, creating a new version of a resource, deleting a resource, etc.

File: theme/templates/resource-landing-page/top-right-buttons.html (hunk 17 above)
Comment (Maurier) [reply theme/templates/resource-landing-page/top-right-buttons.html]: Shareable is addressed here and has been moved to the Vue instance:
https://github.com/hydroshare/hydroshare/blob/3186-manage-access-vue-instance/theme/templates/resource-landing-page/manage-access.html#L207

File: theme/templates/resource-landing-page/subject.html (hunk 18 above)
Comment (Maurier) [reply theme/templates/resource-landing-page/subject.html]: I did. Will fix it.

File: theme/templates/resource-landing-page/quota.html (hunk 20 above)
Comment (Maurier) [reply theme/templates/resource-landing-page/quota.html]: Yes, I'll leave it in place until the next PR with the reusable component is merged. This is addressed there.

File: hs_core/models.py (hunk 2 above)
Comment (Maurier) [reply hs_core/models.py]: @sblack-usu - I added the method as you suggested and I think it improved readability. Please have another look.

**Comment by Maurier** [comment]
retest this, please

File: hs_core/templates/pages/collaborate.html (hunk 13 above)
Comment (cuahsimarko) [reply hs_core/templates/pages/collaborate.html]: If memory serves, we are generally sometimes doing conditional rendering, using Django template language and previously sometimes with JavaScript/jQuery (and now Vue). Ideally, we'll start to lean towards Vue and have Django conditional rendering of DOM/HTML only in cases where Vue doesn't make sense or literally can't do it (access to Python scope impossible or laborious, hard to read code)

File: hs_core/templates/pages/genericresource.html (hunk 14 above)
Comment (cuahsimarko) [reply hs_core/templates/pages/genericresource.html]: Hopefully, we can aim to have less entangled code and more clear handoffs between contexts Django template/HTML/CSS/DOM .... JavaScript/Vue templates/components .... Python/backend/API

**Comment by Maurier** [comment]
@cuahsimarko - The bug with the close button has been fixed. I believe I have addressed @sblack-usu's suggestions.
